### PR TITLE
Add support to raise critical events on thread failures

### PR DESCRIPTION
**Pull request by fruch**
the feature that was reverted (#1361)

(cherry picked from commit 4348223651e1ac9fe5d24fdbd58b6904d90be7ab)

## PR pre-checks (self review)

- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I gave variables/functions meaningful self-explanatory names
- [ ] I didn't leave commented-out/debugging code
- [ ] I didn't copy-paste code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

File: sdcm/utils/thread.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    """
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        result = None
+        _test_pid = os.getpid()
+        from sdcm.cluster import Setup
+        from sdcm.sct_events import ThreadFailedEvent
+        try:
+            result = func(*args, **kwargs)
+        except AssertionError as ex:
```
Comment (bentsi) [inline sdcm/utils/thread.py]: remove this

Comment (fruch) [reply sdcm/utils/thread.py]: fixed

File: sdcm/utils/thread.py [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        result = None
+        _test_pid = os.getpid()
+        from sdcm.cluster import Setup
+        from sdcm.sct_events import ThreadFailedEvent
+        try:
+            result = func(*args, **kwargs)
+        except Exception as ex:  # pylint: disable=broad-except
+            Setup.tester_obj().result.addFailure(Setup.tester_obj(), sys.exc_info())
```
Comment (bentsi) [inline sdcm/utils/thread.py]: Why do you add a failure to the tester object?

Comment (fruch) [reply sdcm/utils/thread.py]: In the previous implementation that was using signal to stop the test immediately, the exception was add to the test as failure or error. 

this mean that when the test fail, you would see this exception as part of the reason the test failed.

I think it's adding a bit more clarity when reading the logs, but also having only the event might be good enough.

Comment (bentsi) [reply sdcm/utils/thread.py]: let's remove this and use Event only (if we want to fail the test we can raise the severity to critical, but let's keep it as error)

Comment (fruch) [reply sdcm/utils/thread.py]: I think both Error and Critical fail the test.

But I'll remove it anyhow

File: sdcm/cluster.py [hunk 3]
```diff
@@ -196,6 +198,15 @@ def set_test_id(cls, test_id):
         else:
             LOGGER.warning("TestID already set!")
 
+    @classmethod
+    def tester_obj(cls):
```
Comment (bentsi) [inline sdcm/cluster.py]: it is not used anywhere in this PR. Is there any reason we need it?

File: sdcm/tester.py [hunk 4]
```diff
@@ -203,6 +203,11 @@ def __init__(self, methodName='test'):  # pylint: disable=too-many-statements
         time.sleep(0.5)
         InfoEvent('TEST_START test_id=%s' % cluster.Setup.test_id())
 
+    def run(self, result=None):
```
Comment (bentsi) [inline sdcm/tester.py]: Why did you add this?

File: sdcm/cluster.py (hunk 3 above)
Comment (fruch) [reply sdcm/cluster.py]: We add for having global way to get to the tester object to add failures or error. (I was in use until this morning :))

File: sdcm/tester.py (hunk 4 above)
Comment (fruch) [reply sdcm/tester.py]: again, it was for having a way to add failures/errors into the tester object

File: sdcm/cluster.py (hunk 3 above)
Comment (bentsi) [reply sdcm/cluster.py]: so please remove this

File: sdcm/cluster.py [hunk 5]
```diff
@@ -181,6 +182,7 @@ class Setup():
     _test_name = None
     TAGS = dict()
     _logdir = None
+    _tester_obj = None
```
Comment (bentsi) [inline sdcm/cluster.py]: remove unused attr
